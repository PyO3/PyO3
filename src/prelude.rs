//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
