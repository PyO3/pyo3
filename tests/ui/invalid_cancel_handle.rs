use pyo3::prelude::*;

#[pyfunction]
async fn cancel_handle_repeated(#[pyo3(cancel_handle, cancel_handle)] _param: String) {}

#[pyfunction]
async fn cancel_handle_repeated2(
    #[pyo3(cancel_handle)] _param: String,
    #[pyo3(cancel_handle)] _param2: String,
) {
}

#[pyfunction]
fn cancel_handle_synchronous(#[pyo3(cancel_handle)] _param: String) {}

#[pyfunction]
async fn cancel_handle_wrong_type(#[pyo3(cancel_handle)] _param: String) {}

#[pyfunction]
async fn missing_cancel_handle_attribute(_param: pyo3::coroutine::CancelHandle) {}

#[pyfunction]
async fn cancel_handle_and_from_py_with(
    #[pyo3(cancel_handle, from_py_with = cancel_handle)] _param: pyo3::coroutine::CancelHandle,
) {
}

fn main() {}
