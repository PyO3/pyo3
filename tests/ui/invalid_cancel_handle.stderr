error: `cancel_handle` may only be specified once per argument
 --> tests/ui/invalid_cancel_handle.rs:4:55
  |
4 | async fn cancel_handle_repeated(#[pyo3(cancel_handle, cancel_handle)] _param: String) {}
  |                                                       ^^^^^^^^^^^^^

error: `cancel_handle` may only be specified once
 --> tests/ui/invalid_cancel_handle.rs:9:28
  |
9 |     #[pyo3(cancel_handle)] _param2: String,
  |                            ^^^^^^^

error: `cancel_handle` attribute can only be used with `async fn`
  --> tests/ui/invalid_cancel_handle.rs:14:53
   |
14 | fn cancel_handle_synchronous(#[pyo3(cancel_handle)] _param: String) {}
   |                                                     ^^^^^^

error: `from_py_with` and `cancel_handle` cannot be specified together
  --> tests/ui/invalid_cancel_handle.rs:24:12
   |
24 |     #[pyo3(cancel_handle, from_py_with = "cancel_handle")] _param: pyo3::coroutine::CancelHandle,
   |            ^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/invalid_cancel_handle.rs:16:1
   |
16 | #[pyfunction]
   | ^^^^^^^^^^^^^
   | |
   | expected `String`, found `CancelHandle`
   | arguments to this function are incorrect
   |
note: function defined here
  --> tests/ui/invalid_cancel_handle.rs:17:10
   |
17 | async fn cancel_handle_wrong_type(#[pyo3(cancel_handle)] _param: String) {}
   |          ^^^^^^^^^^^^^^^^^^^^^^^^                        --------------
   = note: this error originates in the attribute macro `pyfunction` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `CancelHandle: PyFunctionArgument<'_, '_, false>` is not satisfied
  --> tests/ui/invalid_cancel_handle.rs:20:50
   |
20 | async fn missing_cancel_handle_attribute(_param: pyo3::coroutine::CancelHandle) {}
   |                                                  ^^^^ the trait `PyClass` is not implemented for `CancelHandle`
   |
   = help: the trait `PyClass` is implemented for `pyo3::coroutine::Coroutine`
   = note: required for `CancelHandle` to implement `FromPyObject<'_>`
   = note: required for `CancelHandle` to implement `FromPyObjectBound<'_, '_>`
   = note: required for `CancelHandle` to implement `PyFunctionArgument<'_, '_, false>`
note: required by a bound in `extract_argument`
  --> src/impl_/extract_argument.rs
   |
   | pub fn extract_argument<'a, 'py, T, const IS_OPTION: bool>(
   |        ---------------- required by a bound in this function
...
   |     T: PyFunctionArgument<'a, 'py, IS_OPTION>,
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `extract_argument`

error[E0277]: the trait bound `CancelHandle: PyFunctionArgument<'_, '_, false>` is not satisfied
  --> tests/ui/invalid_cancel_handle.rs:20:50
   |
20 | async fn missing_cancel_handle_attribute(_param: pyo3::coroutine::CancelHandle) {}
   |                                                  ^^^^ the trait `Clone` is not implemented for `CancelHandle`
   |
   = help: the following other types implement trait `PyFunctionArgument<'a, 'py, IS_OPTION>`:
             `&'a mut pyo3::coroutine::Coroutine` implements `PyFunctionArgument<'a, 'py, false>`
             `&'a pyo3::Bound<'py, T>` implements `PyFunctionArgument<'a, 'py, false>`
             `&'a pyo3::coroutine::Coroutine` implements `PyFunctionArgument<'a, 'py, false>`
             `Option<T>` implements `PyFunctionArgument<'a, 'py, true>`
   = note: required for `CancelHandle` to implement `FromPyObject<'_>`
   = note: required for `CancelHandle` to implement `FromPyObjectBound<'_, '_>`
   = note: required for `CancelHandle` to implement `PyFunctionArgument<'_, '_, false>`
note: required by a bound in `extract_argument`
  --> src/impl_/extract_argument.rs
   |
   | pub fn extract_argument<'a, 'py, T, const IS_OPTION: bool>(
   |        ---------------- required by a bound in this function
...
   |     T: PyFunctionArgument<'a, 'py, IS_OPTION>,
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `extract_argument`
